visualizer/timeline.js

```diff
@@ -139,35 +139,89 @@ TimelineLayout.prototype._calcBackgroundLine = function (node) {
          `H${this._xScale(flatten.total)}`; // Horizontal line to
 };
 
-TimelineLayout.prototype._calcWaitLine = function (node) {
-  const path = [];
-
-  // Between (init - before) and (after - before)
-  let prevTime = node.init;
-  for (let i = 0; i < node.before.length; i++) {
-    path.push(`M${this._xScale(prevTime)} ${node.top * timelineHeight} ` + // Move to
-              `H${this._xScale(node.before[i])}`); // Horizontal line to
-    prevTime = node.after[i];
-  }
+TimelineLayout.prototype._calcStateLines = function (self) {
+
+  // Wrap our d3 callback so we can pass our instance as "self"
+  return function(node) {
+    // Get our bar element
+    const path = d3.select(this)
+
+    // Create lists to populate path segments
+    const wait = []
+    const wait_unref = []
+    const callback = []
+
+    let prevTime = node.init;
+    // Set the initial unref value
+    let unrefed = node._unrefed;
+    let inCallback = false;
+
+    // Iterate over the state changes
+    for (const state of node.stateChanges) {
+      let endPoint = null;
+      let pathPart = null;
+
+      // Check the stat change and grab both the endPoint time
+      // and which path we should be drawing to
+      if (state[1] === 'unref') {
+        if (unrefed) continue;
+        unrefed = true;
+        if (inCallback) continue;
+
+        endPoint = state[0];
+        pathPart = wait;
+      }
+      if (state[1] === 'ref') {
+        if (!unrefed) continue;
+        unrefed = false;
+        if (inCallback) continue;
 
-  // Between (init/after - destroy)
-  path.push(`M${this._xScale(prevTime)} ${node.top * timelineHeight} ` + // Move to
-            `H${this._xScale(node.destroy)}`); // Horizontal line to
+        endPoint = state[0];
+        pathPart = wait_unref;
+      }
+      if (state[1] === 'before') {
+        inCallback = true;
+        endPoint = state[0];
+        if (unrefed) {
+          pathPart = wait_unref;
+        } else {
+          pathPart = wait;
+        }
+      }
+      if (state[1] === 'after') {
+        inCallback = false;
+        endPoint = state[0];
+        pathPart = callback;
+      }
+      if (endPoint === null ||
```

Comment:
Lets not prepare for the future :p – and in that hypothetical future I think `if { ... } else if { ... } else { continue; }` would be simpler.